### PR TITLE
[KOGITO-8434] Add Integration Tests and configure Container Builder CI

### PR DESCRIPTION
**Pull request by desmax74**
See:https://issues.redhat.com/browse/KOGITO-8434

Signed-off-by: max <mdessi@redhat.com>

**Description of the change:**

**Motivation for the change:**

**Checklist**

- [ ] Add or Modify a unit test for your change
- [ ] Have you verified that tall the tests are passing?

### CODE REVIEW

**Comment by desmax74** [comment]
jenkins test

**Comment by ricardozanini** [comment]
Please wait for https://github.com/kiegroup/container-builder/pull/15 to be merged.

File: .github/workflows/test-go.yml [hunk 1]
```diff
@@ -24,4 +28,8 @@ jobs:
       - run: make test
       - uses: shogo82148/actions-goveralls@v1
         with:
-          path-to-profile: cover.out
\ No newline at end of file
+          path-to-profile: cover.out
+      - name: Docker Integration test
+        run:  make docker-integration-test
+      - name: Podman Integration test
+        run:  make podman-integration-test
```
Comment (radtriste) [inline .github/workflows/test-go.yml]: I would create different jobs for docker and podman testing

File: .github/workflows/test-go.yml [hunk 2]
```diff
@@ -14,7 +14,11 @@ jobs:
           run: |
           sudo apt-get -y install
           btrfs-progs \
-          libgpgme-dev
+          libgpgme-dev \
```
Comment (spolti) [inline .github/workflows/test-go.yml]: please align it with the changes that @ricardozanini is doing.

Comment (ricardozanini) [reply .github/workflows/test-go.yml]: This will be overridden by the other PR.

**Comment by desmax74** [comment]
jenkins test

File: .github/workflows/test-go.yml [hunk 3]
```diff
@@ -6,10 +6,12 @@ on:
       - main
 
 jobs:
-  unit:
-    name: unit
-    runs-on: ubuntu-20.04
-    steps:
+  jobInstall:
```
Comment (radtriste) [inline .github/workflows/test-go.yml]: why not using a matrix strategy ?

Comment (desmax74) [reply .github/workflows/test-go.yml:9]: I've found the matrix strategy to use different version of go for the tests, how can be used to run different jobs with different tests ?

File: cleaner/docker_integration_test.go [hunk 4]
```diff
@@ -19,65 +19,53 @@
...
+	}
+
+	assert.Nil(suite.T(), tagErr, "Tag image failed")
+	time.Sleep(2 * time.Second) // Needed on CI
+	pushErr := suite.Docker.PushImage(TEST_IMAGE_LOCAL_TAG, REGISTRY_CONTAINER_URL_FROM_DOCKER_SOCKET, "", "")
+	if pushErr != nil {
+		logrus.Infof("Push Error:%s", pushErr)
+	}
+
+	assert.Nil(suite.T(), pushErr, "Push image in the DOcker container failed")
```
Comment (davidesalerno) [inline cleaner/docker_integration_test.go]:
```suggestion
	assert.Nil(suite.T(), pushErr, "Push image in the Docker container failed")
```

File: cleaner/docker_integration_test_suite.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+func (suite *DockerTestSuite) TearDownSuite() {
+	registryID := suite.LocalRegistry.GetRegistryRunningID()
+	if len(registryID) > 0 {
+		DockerTearDown(suite.LocalRegistry)
+	} else {
+		suite.LocalRegistry.StopRegistry()
+	}
+	purged, err := suite.Docker.PurgeContainer("", REGISTRY)
+	if err != nil {
+		logrus.Errorf("Error during purged container in TearDown SUite %t", err)
```
Comment (davidesalerno) [inline cleaner/docker_integration_test_suite.go]:
```suggestion
		logrus.Errorf("Error during purged container in TearDown Suite %t", err)
```

File: cleaner/registry_docker.go [hunk 6]
```diff
@@ -176,22 +178,47 @@ func (d DockerLocalRegistry) IsRegistryImagePresent() bool {
...
 	dockerSock := Docker{Connection: dockerSocketConn}
-	dockerSock.PurgeContainer("", REGISTRY)
+	logrus.Infof("Pull images for test")
+	err = dockerSock.PullImage(REGISTRY_FULLWITH_TAG)
+	if err != nil {
+		logrus.Errorf("Pull error in SetupDockerSocket %s", err)
+	}
+	_, err = dockerSock.PurgeContainer("", REGISTRY)
+	if err != nil {
+		logrus.Errorf("Purge  error in SetupDockerSocket %s", err)
```
Comment (davidesalerno) [inline cleaner/registry_docker.go:207]:
```suggestion
		logrus.Errorf("Purge error in SetupDockerSocket %s", err)
```

File: cleaner/docker_integration_test_suite.go (hunk 5 above)
Comment (desmax74) [reply cleaner/docker_integration_test_suite.go]: fixed

File: cleaner/registry_docker.go (hunk 6 above)
Comment (desmax74) [reply cleaner/registry_docker.go:207]: fixed

File: cleaner/docker_integration_test.go (hunk 4 above)
Comment (desmax74) [reply cleaner/docker_integration_test.go]: fixed

File: cleaner/docker_integration_test.go [hunk 7]
```diff
@@ -19,65 +19,53 @@
...
-	if len(registryID) > 0 {
-		DockerTearDown(suite.LocalRegistry)
-	} else {
-		suite.LocalRegistry.StopRegistry()
-	}
-	purged, _ := suite.Docker.PurgeContainer("", REGISTRY)
-	logrus.Infof("Purged containers %t", purged)
-}
-
 // --------------------------- TESTS -----------------
```
Comment (spolti) [inline cleaner/docker_integration_test.go]:
```suggestion
```

File: cleaner/docker_integration_test_suite.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package cleaner
+
+import (
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+// --------------------------- TEST SUITE -----------------
```
Comment (spolti) [inline cleaner/docker_integration_test_suite.go]:
```suggestion
```

File: cleaner/podman_integration_test.go [hunk 9]
```diff
@@ -18,46 +18,18 @@
...
-	registryID := suite.LocalRegistry.GetRegistryRunningID()
-	if len(registryID) > 0 {
-		PodmanTearDown(suite.LocalRegistry)
-	} else {
-		suite.LocalRegistry.StopRegistry()
-	}
-	suite.Podman.PurgeContainer("", REGISTRY_FULL)
-}
-
 // --------------------------- TESTS -----------------
```
Comment (spolti) [inline cleaner/podman_integration_test.go]:
```suggestion
```

File: cleaner/podman_integration_test_suite.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ * limitations under the License.
+ */
+package cleaner
+
+import (
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+// --------------------------- TEST SUITE -----------------
```
Comment (spolti) [inline cleaner/podman_integration_test_suite.go]:
```suggestion
```

File: cleaner/registry_docker.go [hunk 11]
```diff
@@ -72,8 +72,9 @@ func (d DockerLocalRegistry) StartRegistry() string {
 
 	ctx := context.Background()
 	registryID := d.GetRegistryRunningID()
-	logrus.Infof("registryID %s", registryID)
+
 	if len(registryID) > 0 {
+		logrus.Infof("registryID %s", registryID)
 		logrus.Info("Registry already running")
```
Comment (spolti) [inline cleaner/registry_docker.go]:
```suggestion
```

File: cleaner/registry_docker.go [hunk 12]
```diff
@@ -72,8 +72,9 @@ func (d DockerLocalRegistry) StartRegistry() string {
 
 	ctx := context.Background()
 	registryID := d.GetRegistryRunningID()
-	logrus.Infof("registryID %s", registryID)
+
 	if len(registryID) > 0 {
+		logrus.Infof("registryID %s", registryID)
```
Comment (spolti) [inline cleaner/registry_docker.go]:
```suggestion
		logrus.Infof("Registry ID %s is already running", registryID)
```

File: cleaner/registry_docker.go [hunk 13]
```diff
@@ -88,6 +89,7 @@ func (d DockerLocalRegistry) StartRegistry() string {
 	} else {
 		logrus.Info("Registry Image ready")
```
Comment (spolti) [inline cleaner/registry_docker.go]:
```suggestion
		logrus.Info("Creating container for %", imageName )
```

File: cleaner/registry_docker.go [hunk 14]
```diff
@@ -88,6 +89,7 @@ func (d DockerLocalRegistry) StartRegistry() string {
...
 	}
+	time.Sleep(2 * time.Second) // needed on CI
 
 	logrus.Info("Registry Container Create")
 	resp, err := d.Connection.ContainerCreate(ctx, &container.Config{
```
Comment (spolti) [inline cleaner/registry_docker.go:93]: what is the value of REGISTRY env?
shouldn't it be renamed?

File: cleaner/registry_docker.go [hunk 15]
```diff
@@ -176,22 +178,47 @@ func (d DockerLocalRegistry) IsRegistryImagePresent() bool {
...
+	_, err = dockerSock.PurgeContainer("", REGISTRY)
+	if err != nil {
+		logrus.Errorf("Purge  error in SetupDockerSocket %s", err)
+	}
 
 	d := DockerLocalRegistry{Connection: dockerSocketConn}
+	logrus.Info("Check if registry image is present :", d.IsRegistryImagePresent())
 	if !d.IsRegistryImagePresent() {
-		dockerSock.PullImage(TEST_IMAGE_TAG)
+		dockerSock.PullImage(REGISTRY_FULLWITH_TAG)
```
Comment (spolti) [inline cleaner/registry_docker.go]: REGISTRY or IMAGE?

File: cleaner/registry_docker.go [hunk 16]
```diff
@@ -176,22 +178,47 @@ func (d DockerLocalRegistry) IsRegistryImagePresent() bool {
...
 	}
 	registryID := d.GetRegistryRunningID()
 	if len(registryID) == 0 {
```
Comment (spolti) [inline cleaner/registry_docker.go:216]: there are another place using a very similar `if, can't it be moved to a function to be reused?

File: cleaner/registry_docker_integration_test.go [hunk 17]
```diff
@@ -19,99 +19,66 @@
...
-	"time"
-
 	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/suite"
+	"testing"
+	"time"
 )
 
 // --------------------------- TEST SUITE -----------------
```
Comment (spolti) [inline cleaner/registry_docker_integration_test.go]:
```suggestion
```

File: cleaner/registry_podman_integration_test.go [hunk 18]
```diff
@@ -28,86 +28,61 @@ import (
 )
 
 // --------------------------- TEST SUITE -----------------
```
Comment (spolti) [inline cleaner/registry_podman_integration_test.go]:
```suggestion
```

File: cleaner/registry_podman_integration_test.go [hunk 19]
```diff
@@ -28,86 +28,61 @@ import (
...
-	if len(registryID) > 0 {
-		PodmanTearDown(suite.LocalRegistry)
-	} else {
-		suite.LocalRegistry.StopRegistry()
-	}
-	suite.Podman.PurgeContainer("", REGISTRY_FULL)
+	suite.Run(t, new(PodmanTestSuite))
 }
 
 // -------------------------------------- TESTS -----------------------------
```
Comment (spolti) [inline cleaner/registry_podman_integration_test.go]:
```suggestion
```

File: cleaner/registry_docker.go (hunk 13 above)
Comment (desmax74) [reply cleaner/registry_docker.go]: Wrong suggestion ?

File: cleaner/registry_docker.go (hunk 14 above)
Comment (desmax74) [reply cleaner/registry_docker.go:93]: "registry", renamed key into REGISTRY_IMG

File: cleaner/registry_docker.go (hunk 15 above)
Comment (desmax74) [reply cleaner/registry_docker.go]: renamed into REGISTRY_IMG_FULL_TAG

File: cleaner/registry_docker.go (hunk 16 above)
Comment (desmax74) [reply cleaner/registry_docker.go:216]: ok

File: .github/workflows/test-go.yml (hunk 1 above)
Comment (desmax74) [reply .github/workflows/test-go.yml]: Done

File: .github/workflows/test-go.yml (hunk 2 above)
Comment (desmax74) [reply .github/workflows/test-go.yml]: Done

**Review by radtriste — APPROVED** [review]

**Review by spolti — APPROVED** [review]

**Review by davidesalerno — APPROVED** [review]

File: .github/workflows/test-go.yml [hunk 20]
```diff
@@ -1,27 +1,76 @@
...
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Install packages
+        uses: awalsh128/cache-apt-pkgs-action@latest
         with:
-          path-to-profile: cover.out
\ No newline at end of file
+          packages: libbtrfs-dev libgpgme-dev libdevmapper-dev
+          version: 1.0
+      - name: Install Podman 4.3.1
```
Comment (ricardozanini) [inline .github/workflows/test-go.yml:56]: We could have a podman installer action. Of course a future project. :)

**Review by ricardozanini — APPROVED** [review]

**Pull request merged**